File: solutions/number-of-provinces/dfs.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length, provinces = 0;
        boolean[] seen = new boolean[n];
        for (int c = 0; c < n; c++) {
            if (seen[c]) continue;
            provinces++;
            Deque<Integer> stack = new ArrayDeque<>();
            stack.push(c);
            seen[c] = true;
            while (!stack.isEmpty()) {
                int x = stack.pop();
                for (int j = 0; j < n; j++)
                    if (isConnected[x][j] == 1 && !seen[j]) { seen[j] = true; stack.push(j); }
            }
        }
        return provinces;
    }
}

File: solutions/number-of-provinces/dfs.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        seen = [False] * n
        provinces = 0
        for c in range(n):
            if seen[c]:
                continue
            provinces += 1
            seen[c] = True
            stack = [c]
            while stack:
                x = stack.pop()
                for j in range(n):
                    if isConnected[x][j] and not seen[j]:
                        seen[j] = True
                        stack.append(j)
        return provinces
